mlflow/models/utils.py

```diff
@@ -308,10 +321,14 @@ def _enforce_tensor_spec(
                     actual_shape, expected_shape
                 )
             )
-    if clean_tensor_type(actual_type) != tensor_spec.type:
+    # Second clause provides support for signatures which support ragged arrays, but where the
+    # actual input isn't ragged
+    if clean_tensor_type(actual_type) != expected_type and (
+        -1 not in expected_shape[1:] or expected_type != np.dtype("O")
+    ):
```

Comment:
Does this clause cover anything that isn't covered by both clauses above?

Reply:
Yeah, so the the two clauses above handle the default inferred signature, and when the input is a ragged array. If the input isn't a ragged array and a more manual signature has been provided then we want to do more of the normal checking (like the shape check for loop). However, there's a case where the signature provided is an object type, for a ragged array, but the input is a standard type, from a standard array, and we don't want to fail on that (this is a subset of the example that started this conversation).

I thought about including this more above (basically, if the signature has -1s at all just pass the input along, but figured we'd want to do some validation if at all possible.

Reply:
Doesn't the first clause above handle the case where the signature is an object type for a ragged array?

```
if len(expected_shape) == 1 and expected_shape[0] == -1 and expected_type == np.dtype("O")
```

Let me know if I'm missing something here.

Reply:
Not quite, I'm talking about when the signature is more manually provided, but still provided as an object type. So spec == TensorSpec('object', (-1, -1, 2, 3) or something like that, and input is just a standard numpy array, shape == (2, 3, 2, 3), type == int64.

So the idea is, since the spec is more detailed we want to compare shapes as best as possible, using the loop on 317, and then compare type as well unless users just provide it as expecting an object, at which point we just say "okay, we did our best".

Let me know if it's still unclear or if you think we should handle it differently

Reply:
Got it. Can you provide an example input with type `TensorSpec('object', (-1, -1, 2, 3))` that we need to support?

Reply:
Sorry, I'm not sure I understand what you mean. Since that's just some manual defined signature, that's the user saying that they're expecting a batch of objects where the first dimension can vary, but the second and third dimensions are expected to be 2 and 3 respectively. and then the actual input is an array which isn't ragged, so it ends up with a normal shape, say (1, 2, 2, 3), and a normal type, say np.int64.

Reply:
I can't say I know of a use case where if the user is manually providing a signature that they would still specify that it takes an object, but perhaps there's some case where they'd support float to int casting or something like that so they could take both, and so they just put object in the signature? I'm not sure, but it seemed better to air on the side of broader support for it.

Reply:
Got it. Yeah, your example differs from the previous comment about the expected data type being 'object' (float64 != object). I'd err on the side of excluding multi-dimensional object inputs from the scope here, unless there's strong motivation. Just not sure I see the use case.

Reply:
I mean, the expected data type being object and not matching the input is only because of this weird ragged array scenario, anything else and it makes more sense to match exactly, imo. But I can remove this

Reply:
@dbczumar one more follow up here, because this still feels weird to me. The second check I added above works for the scenario I outline in the comments:
# Sample spec: Tensor('object', (-1, -1, -1, 3))
# Will pass on inputs which are ragged arrays: shape==(x,), dtype=='object'

However, it feels weird then that that same signature will fail when the input isn't actually ragged, just because numpy can do better at inferring type at that point. A signature like that shouldn't _require_ that the input be ragged, it's just effectively saying it can be, right?

If you're saying that the first clause I added, the inferred signature handling, is what users should use if they can handle both ragged and non-ragged arrays I guess that's fine, but if that's true then I feel like we should remove the second clause all together and just not support custom signatures at all.

Reply:
The first and second check makes perfect sense to me. It's the third check that I don't fully understand. `Tensor('object', (-1, -1, -1, 3))` doesn't seem like a valid use case, but `Tensor('float64', (-1, -1, -1, 3))` does. For `Tensor('float64', (-1, -1, -1, 3))`, if the input is not ragged, it should already succeed on `master`; if the input is ragged, the second case will handle it.

Reply:
Unless I'm missing something, I'd propose we remove the third check since `Tensor('object', (-1, -1, -1, 3))` isn't a conventionally valid signature and also change the example on Line 309 to read `Tensor('float64', (-1, -1, -1, 3))`.

Reply:
Considering the expanded signature with multiple -1 isn't a conventionally valid signature either I'm not sure I agree with the argument, but don't think it's worth drawing it out. I've removed the check, if we get feedback that people want an object type there then we can add it back.